src/parser/gate.rs

```diff
@@ -22,18 +22,19 @@ use super::{
     ParserInput, ParserResult,
 };
 use crate::instruction::Gate;
+use crate::parser::error::ignore_leftover_error;
 
 /// Parse a gate instruction.
 pub fn parse_gate<'a>(input: ParserInput<'a>) -> ParserResult<'a, Instruction> {
-    let (input, modifiers) = many0(parse_gate_modifier)(input)?;
+    let (input, modifiers) = many0(ignore_leftover_error(parse_gate_modifier))(input)?;
     let (input, name) = token!(Identifier(v))(input)?;
     let (input, parameters) = opt(delimited(
         token!(LParenthesis),
-        separated_list0(token!(Comma), parse_expression),
+        separated_list0(token!(Comma), ignore_leftover_error(parse_expression)),
         token!(RParenthesis),
     ))(input)?;
     let parameters = parameters.unwrap_or_default();
-    let (input, qubits) = many0(common::parse_qubit)(input)?;
+    let (input, qubits) = many0(ignore_leftover_error(common::parse_qubit))(input)?;
```

Comment:
Overall, the need for such wide use of `ignore_leftover_error` isn't clear to me. In this particular case, for example, `parse_qubit` can't return a `LeftoverError` variant anyway, and an error for that wouldn't make sense in the conventions of `nom` anyway (erroring on leftover is more an opt-in thing with `all_consuming` but not really applicable elsewhere).

The fact that there are panics to handle this and that there are ~100 invocations of this parser modifier suggests to me that there's a neater way that doesn't make a leftover error an edge case that needs special handling everywhere. I'd probably have to spend some time hands-on with it, either solo or alongside you, to better understand what that might look like.

Reply:
If it helps, here's an explanation of why `ignore_leftover_error` exists.

`(Parse)Error` (hereafter `Error`) contains a variant `Leftover` that contains a `LeftoverError { leftover: I, parsed: O }`. This allows consumers to optionally ignore that the input wasn't fully consumed and use the parsed value anyways.

This is stored in `Error` because it can be returned along with the other error variants. Because the input and returned types can change (and because sometimes other errors can occur), `Error` is defined as `Error<I, O, E>`.

`ignore_leftover_error` and `Error::convert_not_leftover` convert `Error<I, O, E>` into an `Error<I, O2, E>`. _This_ is necessary because if `parser1(I) -> IResult<I, O2, Error<I, O2, E>>` uses `parser2(I) -> IResult<I, O, Error<I, O, E>>`, the error type `O2` needs to be changed to `O`, which is essentially free if `Error::Leftover` is impossible.

----------

That said, taking a second look at this, it looks possible that `LeftoverError` can be moved from `Error` to `ProgramError`. I thought it was used in at least one place where `ProgramError` _wasn't_ used, but I couldn't find it through quick search through the code.

I'll take a look tomorrow at trying the refactor. If it works, that will simplify a lot of these changes.

Reply:
Removed in recent commits.